### PR TITLE
perf(pnp): optimize `findPackageLocator`

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

`findPackageLocator` is used by every function in the `pnpapi` and could be faster

**How did you fix it?**

Removed the `packageLocationLengths` logic and replaced it with `lastIndexOf` to check the sections of the filepath

**Result**

Time spent in `findPackageLocator`
```diff
# yarn build:doc
- 36472 calls, 541.3566000048304 ms
+ 36294 calls, 472.3983999992197 ms

# Webpack build from a private project with enableGlobalCache: true
- 27247 calls, 631.8915001358837 ms
+ 27255 calls, 439.93869973160326 ms
```

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-pnp/sources/loader/makeApi.ts [hunk 1]
```diff
@@ -497,22 +496,16 @@ export function makeApi(runtimeState: RuntimeState, opts: MakeApiOptions): PnpAp
     if (!relativeLocation.endsWith(`/`))
       relativeLocation = `${relativeLocation}/` as PortablePath;
 
-    let from = 0;
+    do {
+      const entry = packageLocatorsByLocations.get(relativeLocation);
 
-    // If someone wants to use a binary search to go from O(n) to O(log n), be my guest
```
Comment (merceyz) [inline packages/yarnpkg-pnp/sources/loader/makeApi.ts:502]: I tested a binary search but wasn't much of a difference

Comment (larixer) [reply packages/yarnpkg-pnp/sources/loader/makeApi.ts:502]: Your binary search is not O(log n) ;-) ?

Comment (merceyz) [reply packages/yarnpkg-pnp/sources/loader/makeApi.ts:502]: I did this back in December but I think I used https://yarnpkg.com/package/binary-search

Comment (merceyz) [reply packages/yarnpkg-pnp/sources/loader/makeApi.ts:502]: Tested again with https://github.com/mikolalysenko/binary-search-bounds

```diff
# yarn bin
- 6609 calls, 107.37779996916652 ms
+ 6609 calls, 122.76100007817149 ms
```

File: packages/yarnpkg-pnp/sources/loader/makeApi.ts [hunk 2]
```diff
@@ -497,22 +496,16 @@ export function makeApi(runtimeState: RuntimeState, opts: MakeApiOptions): PnpAp
...
-    for (let t = from; t < packageLocationLengths.length; ++t) {
-      const entry = packageLocatorsByLocations.get(relativeLocation.substr(0, packageLocationLengths[t]) as PortablePath);
-
-      if (typeof entry !== `undefined`) {
-        if (entry.discardFromLookup && !includeDiscardFromLookup)
-          continue;
-
-        return entry.locator;
+      if (typeof entry === `undefined` || (entry.discardFromLookup && !includeDiscardFromLookup)) {
+        relativeLocation = relativeLocation.substring(0, relativeLocation.lastIndexOf(`/`, relativeLocation.length - 2) + 1) as PortablePath;
```
Comment (arcanis) [inline packages/yarnpkg-pnp/sources/loader/makeApi.ts:503]: Why `n - 2` and not `n - 1`? 🤔

Comment (merceyz) [reply packages/yarnpkg-pnp/sources/loader/makeApi.ts:503]: `n - 1` would only give you the index but we need to skip the trailing slash as well

**Pull request merged**